### PR TITLE
Farewell Workflow Exercise

### PR DESCRIPTION
**Pull request by jsundai**
## Checklist

1. Closes 

2. How was this tested:

3. Any docs updates needed?

### CODE REVIEW

File: exercises/farewell-workflow/README.md [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+you'll need to change. If you get stuck and need additional hints, or 
+if you want to check your work, look at the completed example in the
+`solution` directory. 
+
+## Part A: Write an Activity Function
+The `TranslateActivities.cs` file contains a method(`GetSpanishGreeting`) that uses a 
+microservice to get a customized greeting in Spanish. Follow these steps to create 
+a new Activity that gets a custom farewell message from the microservice:
+
+1. Open the `TranslateActivities.cs` file. This is located in the `practice/Workflow`
```
Comment (MasonEgger) [inline exercises/farewell-workflow/README.md:21]: Just give them the full path. The following sentence makes them have to construct the path in their head
```suggestion
1. Open the `TranslateActivities.cs` file. This is located in the `exercises/farewell-workflow/practice/Workflow`
```

File: exercises/farewell-workflow/README.md [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+   subdirectory under the `farewell-workflow` directory in `exercises`.
```
Comment (MasonEgger) [inline exercises/farewell-workflow/README.md:22]:
```suggestion
   subdirectory
```

File: exercises/farewell-workflow/README.md [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+2. Create a new Activity that will get a custom farewell message.
+   1. Copy the `GetSpanishGreeting` method to create a new Activity.
+   2. Rename the new `GetSpanishGreeting` function to `GetSpanishFarewell`.
```
Comment (MasonEgger) [inline exercises/farewell-workflow/README.md:25]: Add a line here instructing the learner to modify the Activity to call the appropriate URL

```
1. Modify the Activity to call `get-spanish-farewell`
```

File: exercises/farewell-workflow/README.md [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+   3. Save your changes to this file.
+
+## Part B: Register the Activity Function
+1. Open the `Program.cs` file in the Worker subdirectory in your code editor.
+2. Add a new line that registers your new Activity with the Worker (hint: you'll use the name you used for the new method in an earlier step).
+3. Save your changes to this file.
+
+## Part C: Modify the Workflow to Execute Your New Activity
+1. Open the `GreetingWorkflow.cs` file (located in the `practice/Workflow` subdirectory) in the editor.
+2. Look for the TODO message, uncomment the lines below it, and then update this code to execute the Activity function you created.
```
Comment (MasonEgger) [inline exercises/farewell-workflow/README.md:35]: I would include those instructions in here as well, not just in the source file

File: exercises/farewell-workflow/README.md [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+3. Save your changes to this file.
+
+## Part D: Start the Microservice and Run the Workflow
+First, make sure the Temporal Service is running in the background. This is done by opening a new terminal window and running the following command: `temporal server start-dev --ui-port 8080 --db-filename clusterdata.db`. For more details on this command, please refer to the `Setting up a Local Development Environment` chapter in the course. 
+
+Note: If you're using the Gitpod environment to run this exercise, you can skip this step.
```
Comment (MasonEgger) [inline exercises/farewell-workflow/README.md:41]: I would move this to the top so GitPod learners don't accidentally try to do this then realize they didn't need to. The majority of our learners use GitPod

File: exercises/farewell-workflow/README.md [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+
+### Run the Microservice
+1. Open a terminal and navigate to the `practice` subdirectory.
```
Comment (MasonEgger) [inline exercises/farewell-workflow/README.md:44]: Would include the `cd` command as mentioned above.

**Review by MasonEgger — CHANGES_REQUESTED** [review]

File: exercises/farewell-workflow/practice/Web/Web.http [hunk 7]
```diff
@@ -0,0 +1,13 @@
+@Web_HostAddress = http://localhost:5125
+
+### Get Spanish Greeting
```
Comment (axfelix) [inline exercises/farewell-workflow/practice/Web/Web.http:3]: I'm not familiar with `.http` files in dotnet applications but Github's syntax highlighting is suggesting that it doesn't like comments/headers formatted this way... could be wrong and it's fine, just seems worth a head check.

File: exercises/farewell-workflow/solution/Client/Client.sln
Comment (cretz) [inline exercises/farewell-workflow/solution/Client/Client.sln:1]: #5 had a solution for the whole exercise, but this is per project. Is the difference intentional?

File: exercises/farewell-workflow/solution/Workflow/GreetingWorkflow.cs [hunk 8]
```diff
@@ -0,0 +1,25 @@
+namespace Temporalio.Farewell.Workflow;
```
Comment (cretz) [inline exercises/farewell-workflow/solution/Workflow/GreetingWorkflow.cs:1]: #5 used `MyNamespace` but this one uses a more proper namespace. Is the difference intentional? Would like to see #5 get updated with a proper namespace too.

File: exercises/farewell-workflow/solution/Client/Program.cs [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+using Temporalio.Client;
+using Temporalio.Farewell.Workflow;
+
+// Connect to the Temporal server
+var client = await TemporalClient.ConnectAsync(new("localhost:7233") { Namespace = "default" });
+
+var workflowId = $"greeting-{Guid.NewGuid()}";
+
+// Default to "Temporal" and overriding if a valid argument is provided
+string name = (args.Length > 0 && !string.IsNullOrEmpty(args[0])) ? args[0] : "Temporal";  // Default name
```
Comment (cretz) [inline exercises/farewell-workflow/solution/Client/Program.cs:11]:
```suggestion
var name = args.FirstOrDefault() ?? "Temporal";  // Default name
```

This is a bit clearer (untested),. But why default to `"Temporal"` if you're just going to ask later and default again?

File: exercises/farewell-workflow/solution/Client/Program.cs [hunk 10]
```diff
@@ -0,0 +1,39 @@
+﻿// Run Workflow file
+using Temporalio.Client;
+using Temporalio.Farewell.Workflow;
+
+// Connect to the Temporal server
+var client = await TemporalClient.ConnectAsync(new("localhost:7233") { Namespace = "default" });
```
Comment (cretz) [inline exercises/farewell-workflow/solution/Client/Program.cs:6]: Inconsistent with other exercises like #5 on whether you explicitly specify the `"default"` namespace or not

**Review by cretz — COMMENTED** [review]
Only reviewed `solution` and some comments may apply to multiple places but I only made them in the first applicable place. Also comments made on #5 should be assumed to apply here.

File: exercises/farewell-workflow/solution/Client/Program.cs [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+{
+    name = args.FirstOrDefault() ?? "Temporal";
+}
+
+// Start the Workflow
+var result = await client.ExecuteWorkflowAsync(
+    (GreetingWorkflow wf) => wf.RunAsync(name),
+    new(id: workflowId, taskQueue: "farewell-workflow"));
+
+Console.WriteLine($"Workflow result: {result}");  // Output Workflow result
```
Comment (cretz) [inline exercises/farewell-workflow/solution/Client/Program.cs:24]: Inconsistent with whether you use string interpolation or `{0}` like replacers (#5 used the former, this uses the latter)

File: exercises/farewell-workflow/solution/Client/Program.cs [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+
+// Generate a unique Workflow ID
+var workflowId = $"greeting-{Guid.NewGuid()}";
+
+// Prompt the user to enter their name and default to Temporal
+Console.WriteLine("Please enter your name:");
+var name = Console.ReadLine();
+if (string.IsNullOrEmpty(name))
+{
+    name = args.FirstOrDefault() ?? "Temporal";
```
Comment (cretz) [inline exercises/farewell-workflow/solution/Client/Program.cs:16]: Why ask for a user's name at all if it's present in args?

File: exercises/farewell-workflow/solution/Web/Web.csproj [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
+  <PropertyGroup>
+    <TargetFramework>net8.0</TargetFramework>
+    <Nullable>enable</Nullable>
+    <ImplicitUsings>enable</ImplicitUsings>
+    <InvariantGlobalization>true</InvariantGlobalization>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <ProjectReference Include="..\Worker\Worker.csproj" />
+    <ProjectReference Include="..\Workflow\Workflow.csproj" />
```
Comment (cretz) [inline exercises/farewell-workflow/solution/Web/Web.csproj:12]: Are either of these dependencies needed?

File: exercises/farewell-workflow/solution/Web/Web.csproj [hunk 14]
```diff
@@ -0,0 +1,22 @@
...
+  </ItemGroup>
+
+  <ItemGroup>
+    <PackageReference Include="Microsoft.AspNetCore.OpenApi" Version="8.0.0" />
```
Comment (cretz) [inline exercises/farewell-workflow/solution/Web/Web.csproj:16]: Arguably OpenAPI is not needed here

File: exercises/farewell-workflow/solution/Web/Web.csproj [hunk 15]
```diff
@@ -0,0 +1,22 @@
...
+    <PackageReference Include="Microsoft.Extensions.Logging" Version="8.0.0" />
+    <PackageReference Include="Microsoft.Extensions.Logging.Console" Version="8.0.0" />
+    <PackageReference Include="Temporalio" Version="1.1.1" />
```
Comment (cretz) [inline exercises/farewell-workflow/solution/Web/Web.csproj:19]: Does not look like anything Temporal related is done in this project

File: exercises/farewell-workflow/solution/Worker/Program.cs [hunk 16]
```diff
@@ -0,0 +1,20 @@
...
+
+IHost host = Host.CreateDefaultBuilder(args)
+    .ConfigureLogging(ctx =>
+        ctx.AddSimpleConsole().SetMinimumLevel(LogLevel.Information))
+    .ConfigureServices(ctx =>
+        ctx.AddHostedTemporalWorker(
+            clientTargetHost: "localhost:7233",
+            clientNamespace: "default",
+            taskQueue: "farewell-workflow")
+            .AddStaticActivities(typeof(TranslateActivities)) // Add Activities
```
Comment (cretz) [inline exercises/farewell-workflow/solution/Worker/Program.cs:16]:
```suggestion
            .AddStaticActivities<TranslateActivities>() // Add Activities
```

This is a bite more idiomatic

File: exercises/farewell-workflow/solution/Workflow/TranslateActivities.cs [hunk 17]
```diff
@@ -0,0 +1,23 @@
+// Translate Activities
+namespace Temporalio.Farewell.Workflow;
+
+using Temporalio.Activities;
+
+public class TranslateActivities
+{
+    private static readonly HttpClient client = new();
```
Comment (cretz) [inline exercises/farewell-workflow/solution/Workflow/TranslateActivities.cs:8]: There's an opportunity here to use non-static and DI if you want to inject the HTTP client in the constructor, but don't have to

File: exercises/farewell-workflow/solution/Workflow/TranslateActivities.cs [hunk 18]
```diff
@@ -0,0 +1,23 @@
+// Translate Activities
```
Comment (cretz) [inline exercises/farewell-workflow/solution/Workflow/TranslateActivities.cs:1]: Usually no need for leading comments like this in any of the files (but if you want proper .NET doc you _could_, though that's probably unnecessary too)

File: exercises/farewell-workflow/solution/Workflow/TranslateActivities.cs [hunk 19]
```diff
@@ -0,0 +1,23 @@
...
+using Temporalio.Activities;
+
+public class TranslateActivities
+{
+    private static readonly HttpClient client = new();
+
+    [Activity]
+    public static async Task<string> GetSpanishGreeting(string name)
+    {
+        var response = await client.GetAsync($"http://localhost:5125/get-spanish-greeting?name={name}");
```
Comment (cretz) [inline exercises/farewell-workflow/solution/Workflow/TranslateActivities.cs:13]: You need to escape the `name` for use in a URL parameter

**Review by cretz — COMMENTED** [review]
Only reviewed `solution` and some comments may apply to multiple places but I only made them in the first applicable place. Also comments made on https://github.com/temporalio/edu-101-dotnet-code/pull/5 should be assumed to apply here.

Please let me know once all comments are addressed and give me a chance to re-review, thanks!

File: exercises/farewell-workflow/solution/Client/Program.cs [hunk 20]
```diff
@@ -0,0 +1,26 @@
...
+
+if (string.IsNullOrEmpty(name))
+{
+    Console.WriteLine("Please enter your name:");
+    name = Console.ReadLine();
+}
+
+name = string.IsNullOrEmpty(name) ? "Temporal" : name;
+
+// Start the Workflow
```
Comment (cretz) [inline exercises/farewell-workflow/solution/Client/Program.cs:21]:
```suggestion
// Run the Workflow
```

This matches your comments in other PRs

File: exercises/farewell-workflow/solution/Web/Program.cs [hunk 21]
```diff
@@ -0,0 +1,52 @@
...
+var logger = app.Services.GetRequiredService<ILogger<Program>>();
+
+app.MapGet("/get-spanish-greeting", (HttpContext context) =>
+{
+    var query = context.Request.Query["name"];
+    if (query.Count > 0)
+    {
+        var name = query[0];
+        var greeting = $"¡Hola, {name}!";
+        logger.LogInformation($"Returning Spanish greeting: {greeting}");
```
Comment (cretz) [inline exercises/farewell-workflow/solution/Web/Program.cs:27]:
```suggestion
        logger.LogInformation("Returning Spanish greeting: {Greeting}", greeting);
```

This is the more proper way to use a .NET logger (I think analyzers will warn you here)

File: exercises/farewell-workflow/solution/Workflow/TranslateActivities.cs (hunk 17 above)
Comment (cretz) [reply exercises/farewell-workflow/solution/Workflow/TranslateActivities.cs:7]: Bump. This actually may be more important now that I have checked and `HttpClient` is `IDisposable`. I suspect when you turn on analyzers, you will get a warning.

**Review by cretz — COMMENTED** [review]
Looks good, but I think when you turn on analyzers you'll see more things to fix

**Pull request merged**